### PR TITLE
feature/ethers

### LINKED ISSUE #249: Migrate the sdk to `ethers.js`

### LINKED ISSUE #265: Improve handling of txparams for non-eip1559 networks
Improve `ContractBase.send` to better handle to constructions of the transaction parameters based on whether the network supports eip-1559 or not.

### PR DESCRIPTION
**Pull request by r-marques**
## Description

- migration to ethers.js
- removed all web3 related dependencies
- refactored how transaction fees are calculated

## Is this PR related with an open issue?

Resolves #249
Resolves #265 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Checklist:

- [ ] Follows the code style of this project.
- [ ] Tests Cover Changes
- [ ] Documentation

#### Funny gif

Put a link of a funny gif inside the parenthesis-->

### CODE REVIEW

File: integration/nevermined/ContractEvent.test.ts [hunk 1]
```diff
@@ -48,6 +48,7 @@ describe('ContractEvent', () => {
         const waitUntilEvent = new Promise(resolve => {
             subscription = nevermined.keeper.token.events.subscribe(
                 events => {
+                    console.log('--- inside events', events)
```
Comment (aaitor) [inline integration/nevermined/ContractEvent.test.ts:51]: cleanup

File: integration/nevermined/NFTTemplatesWithEther.test.ts [hunk 2]
```diff
@@ -93,8 +93,10 @@ describe('NFTTemplates With Ether E2E', async () => {
         ;({ nftSalesTemplate, nftAccessTemplate } = nevermined.keeper.templates)
 
         // eth
-        nftPrice = Number(web3Utils.toWei(String(nftPrice), 'ether'))
-        amounts = amounts.map(v => new BigNumber(web3Utils.toWei(String(v), 'ether')))
+        nftPrice = ethers.utils.parseEther(nftPrice.toString()).toNumber()
+        amounts = amounts.map(
+            v => new BigNumber(ethers.utils.parseEther(v.toString()).toString())
```
Comment (aaitor) [inline integration/nevermined/NFTTemplatesWithEther.test.ts:98]: weird line, same here, maybe we can isolate this kind of tricky type adaptations in a utility class. That would make easier a further migration to BN.js

File: src/events/ContractEvent.ts [hunk 3]
```diff
@@ -26,16 +27,32 @@ export class ContractEvent extends NeverminedEvent {
...
                 `Event "${options.eventName}" not found on contract "${this.contract.contractName}"`
             )
         }
-        return this.contract.contract.getPastEvents(options.eventName, {
-            filter: options.filterJsonRpc,
-            fromBlock: options.fromBlock,
-            toBlock: options.toBlock
-        })
+        const args = this.filterToArgs(options.eventName, options.filterJsonRpc)
+        console.log('----- ContractEvent getEventData args', args)
```
Comment (aaitor) [inline src/events/ContractEvent.ts:37]: cleanup

File: src/events/ContractEvent.ts [hunk 4]
```diff
@@ -26,16 +27,32 @@ export class ContractEvent extends NeverminedEvent {
     }
 
     public async getEventData(options: EventOptions): EventResult {
-        if (!this.contract.contract.events[options.eventName]) {
+        console.log('----- ContractEvent getEventData')
```
Comment (aaitor) [inline src/events/ContractEvent.ts:30]: cleanup

File: src/events/ContractEvent.ts [hunk 5]
```diff
@@ -47,16 +64,27 @@ export class ContractEvent extends NeverminedEvent {
             // Temporary workaround to work with mumbai
             // Infura as a 1000 blokcs limit on their api
             if (chainId === 80001 || chainId === 42) {
-                const latestBlock = await this.web3.eth.getBlockNumber()
+                const latestBlock = await this.web3.getBlockNumber()
+                console.log('---- ContractEvent latestBlock', latestBlock)
```
Comment (aaitor) [inline src/events/ContractEvent.ts:68]: cleanup

File: src/events/EventHandler.ts [hunk 6]
```diff
@@ -14,23 +14,25 @@ export class EventHandler {
...
 
     private getBlockNumber: () => Promise<number>
-
     constructor() {}
-
     public subscribe(
         callback: (blockNumber: number) => void,
         getBlockNumber: () => Promise<number>
     ) {
+        console.log('--- EventHandler subscribe')
```
Comment (aaitor) [inline src/events/EventHandler.ts:22]: cleanup

File: test/keeper/ContractEvent.test.ts [hunk 7]
```diff
@@ -48,9 +48,13 @@ describe('ContractEvent', () => {
                     }
                 )
             })
+            console.log('before request tokens')
```
Comment (aaitor) [inline test/keeper/ContractEvent.test.ts:51]: cleanup

File: integration/nevermined/Web3Providers.test.ts [hunk 8]
```diff
@@ -6,9 +6,9 @@ import { getMetadata } from '../utils'
 import { Nevermined, Account } from '../../src'
 
 import HDWalletProvider from '@truffle/hdwallet-provider'
```
Comment (mrsmkl) [inline integration/nevermined/Web3Providers.test.ts:8]: Is there a way to get rid of hdwallet-provider?

Comment (r-marques) [reply integration/nevermined/Web3Providers.test.ts:8]: yeah, ethers.js has it's own hdwallet provider

File: integration/nevermined/NFTTemplatesWithEther.test.ts (hunk 2 above)
Comment (r-marques) [reply integration/nevermined/NFTTemplatesWithEther.test.ts:95]: created new issue for that https://github.com/nevermined-io/sdk-js/issues/308

**Review by aaitor — APPROVED** [review]

**Pull request merged**